### DISCUSSION TITLE
Vim Script Support

### DISCUSSION THREAD

**Discussion by AlexPl292** [JetBrains/ideavim#357] [General]
This discussion is about the biggest and most interesting change in IdeaVim - vim script support. 
Related YouTrack Ticket: https://youtrack.jetbrains.com/issue/VIM-506

For some time we've been working on supporting vim script execution. We've completely rewritten our script execution engine so that now IdeaVim builds the full Abstract Syntax Tree and executes it as a real language and not just a list of simple commands.

In the current 1.7.0 release you can find the following state of the vim script support:

- IdeaVim can execute all the instructions that were previously supported.
- Vim script is supported as a language. This includes `if`, `for`, `while` and other commands, data structures (like arrays or dicts), functions execution, and expressions calculation.
- In-built functions are mostly not implemented. This includes more than 600 functions, which we plan to support in the future.  

Here is a full roadmap: https://github.com/JetBrains/ideavim/blob/master/vimscript-info/VIMSCRIPT_ROADMAP.md

For the current release, we aim to support everything that IdeaVim can do now with as few new features as possible; stability and preventing any unexpected behavior are our goals.

Special thanks to @lippfi who did this big and incredible change to the IdeaVim project 🎉

# Examples

  - Simple expression: `:echo 1 + 2`
  - Built-in function: `:echo sin(0.5)`

Using `~/.ideavimrc`:
 - Simple function with branching:
      ```
    function! IsThisFive(number)
        if a:number == 5
            return "This is five"
        else
            return "This is not five"
        endif
    endfunction
    ```

    Then call the function from ex line `:echo IsThisFive(3)`

 - Mapping with condition
     ```
    inoremap <expr> jk col(".") == 1? '<Esc>' : '<Esc><Right>'
    ```

   Follow this link to know how to create a numbered list like a pro: http://vimdoc.sourceforge.net/htmldoc/map.html#:map-expression

# Q&A

- Q: Will I be able to run my plugins on IdeaVim?
- A: Not at the moment, but maybe in the future.
  Running vim script plugins by IdeaVim is one of the goals of vim script support. However, this will require an implementation of built-in functions that are used by this plugin.

--

- Q: How can I play with it?
- A: See `Examples` section.

--

- Q: My configuration is completely broken after this release!
- A: This is a very first release of vim script support, so we expect maybe a few configurations broken. Please share more details about what might be wrong or which command isn’t working for you.

**Comment by BradenM** [comment]
Hi @AlexPl292, 

Super excited about this!

So excited in fact, that I immediately downloaded it... and well...

> Q: My configuration is completely broken after this release!

![gif_out](https://user-images.githubusercontent.com/5913808/129293100-d0cc2ff9-6f67-4b8b-8084-75b44c722848.gif)


It seems (as of `v0.69.1`) -- the whole plugin broke :grimacing: 
I am stuck in "insert" mode and cannot run any commands (essentially acting as if the plugin isn't installed).

My IDE and environment:
```
PyCharm 2021.2 (Professional Edition)
Build #PY-212.4746.96, built on July 27, 2021
Runtime version: 11.0.11+9-b1504.13 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.14.0-1-manjaro
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 20
Registry: ide.experimental.ui.toolwindow.stripes=true, ide.balloon.shadow.size=0
Non-Bundled Plugins: AceJump (3.8.3), IdeaVIM (0.69.1), org.jetbrains.IdeaVim-EasyMotion (1.6), io.github.mishkun.ideavimsneak (1.2.0), Key Promoter X (2021.2), com.intellij.ideolog (203.0.27.0), ideanginx9 (0.1.9), org.jetbrains.plugins.go-template (212.4746.52), zielu.gittoolbox (212.7.6), mobi.hsz.idea.gitignore (4.2.0), name.kropp.intellij.makefile (212.4746.52), com.intellij.kubernetes (212.4746.57), com.theodo.serverless.plugin.serverless-plugin (1.3.4), org.toml.lang (0.2.152.4026-212), ski.chrzanow.foldableprojectview (1.1.1), com.chrisrm.idea.MaterialThemeUI (6.8.0.1), izhangzhihao.rainbow.brackets (6.19), com.intellij.lang.jsgraphql (3.0.0), com.dropbox.plugins.mypy_plugin (0.3.2), com.koxudaxi.poetry (1.1.4-212), com.koxudaxi.pydantic (0.3.4), aws.toolkit (1.30-212), ru.adelf.idea.dotenv (2021.3.0.212)
Current Desktop: sway
```

Other IdeaVim related IntelliJ plugins used:
* IdeaVim-EasyMotion (v1.6)
* IdeaVim-Sneak (v1.2.0)


Here is a gist of my [ideavimrc](https://gist.github.com/BradenM/63deaf9f734d63389731b9f73c16e318)
Can also provide my `vimrc` if requested.

Strangely, I was unable to find any exceptions/crashes thrown by IdeaVim (or related) plugins in my PyCharm logs... but am happy to provide them regardless if requested.


Also, apologies if I misunderstood this discussion board as the place to report/discuss issues. If that's the case, please do redirect me to the appropriate place and I will transfer this report there :+1: . Thanks for the awesome plugin + work!

**Reply by BradenM** [reply]
Ah, seems like I spoke a bit too soon.

The issue persists through restarts, but as you can see in my gif below, it can be resolved (per-session) by disabling+re-enabling the IdeaVim plugin.
However, some of my bindings (namely the common `jk`) appear to be broken.

![gif_out](https://user-images.githubusercontent.com/5913808/129295367-4cad6dd6-d4f5-4c11-91ad-4b74d3293ac0.gif)

**Reply by AlexPl292** [reply]
Hey Braden, thank you for your response!
Glad to hear that you like this feature (expect the bug :) ). We've tried to reproduce the issue with your `~/.ideavimrc` and plugins installed, but it works pretty fine so far. So yes, could you please share your `.vimrc` as well and [logs](https://intellij-support.jetbrains.com/hc/en-us/articles/207241085-Locating-IDE-log-files)? You can leave it here or create a ticket on YouTrack with a restricted visibility, if you like.

---
> Q: My configuration is completely broken after this release!
> A: Hmm, we did not expect that.

Also, sorry for misleading you, actually this is the very first release of vim script support, so we expect maybe a few broken configs, but we are here to help! I'll update the article.

**Reply by AlexPl292** [reply]
Hey, we have found `.vimrc` file in your public dotfiles repo. Our current implementation really fails on it, but we've fixed the bug and now you can try to update the plugin to 0.69.2 EAP.

**Reply by BradenM** [reply]
@AlexPl292 Thanks for taking the time to hunt down my `vimrc`. (I probably should've included it in the first place :grimacing: )

> Also, sorry for misleading you, actually this is the very first release of vim script support, so we expect maybe a few broken configs, but we are here to help!

No worries, I had understood that it's still EAP and was expecting some kinks. Glad to hear you guys resolved the issue.
I'll report back if I run into any issues on the latest version. Thanks!

**Comment by adtran83** [comment]
Hi @AlexPl292,

Thanks for your hard work.

I've updated Ideavim today and I have some problems:
- Space key doesn't behave like leader anymore
- My mappings doesn't work properly, for example, when I do Y, it copy until the end of the line but after that, the cursor is moved to the next line.

This is my idea.vimrc

```
let mapleader = " "

""" Plugins ---------------------------------------------
set easymotion
set surround
set multiple-cursors
set commentary
set ReplaceWithRegister
set argtextobj
set exchange
set textobj-entire
set highlightedyank
set sneak
set NERDTree
set vim-paragraph-motion
set textobj-indent

" Plugins settings --------------------------------------
let g:argtextobj_pairs="(:),{:},[:]"

" Common settings
set visualbell
set noerrorbells
set relativenumber
set number
set ignorecase
set smartcase
set scrolloff=5 " 5 lines above/below cursor when scrolling
set incsearch
set hlsearch
set clipboard+=unnamed
set showmode

" IDEA specific settings
set ideajoin
set ideaput
set ideamarks
set idearefactormode=keep

" Map perso ---------------------------------------------
nnoremap Y y$

nnoremap <C-h> <C-w>h
nnoremap <C-j> <C-w>j
nnoremap <C-k> <C-w>k
nnoremap <C-l> <C-w>l

nnoremap n nzzzv
nnoremap N Nzzzv
nnoremap J mzJ`z

nnoremap cn *Ncgn

```

Thanks,

**Reply by AlexPl292** [reply]
Hi, sorry for the inconvenience :(
We've tried to reproduce any of the described issues, but no luck so far. Could you please share your log (https://intellij-support.jetbrains.com/hc/en-us/articles/207241085-Locating-IDE-log-files) file here or on YouTrack?

**Reply by adtran83** [reply]
Hey, thanks for your help, this is my log file:

[idea.log](https://github.com/JetBrains/ideavim/files/6992255/idea.log)

**Reply by peppy** [reply]
I was also experiencing similar issues. When activating keys, ideavim would move down an extra line before executing the command. Reverting back to the previous release works as a fix for now.

https://drive.google.com/file/d/14-4fbmUaXG7UKYhv8eaDxGYpBcKnD8kq/view?usp=sharing

I was invoking ',' here, which is bound to:

```csharp
nnoremap ,   :action RiderNuGetContextPopupAction<CR>:action ShowIntentionActions<CR>
```

![20210816 214926 (JetBrains Rider)](https://user-images.githubusercontent.com/191335/129566369-40de8aa7-75af-408d-8d80-050093911351.gif)

**Reply by AlexPl292** [reply]
@adtran83 thank you for the logs, we've found out some issues on windows. The fix is on the way, but at the moment you can switch to stable plugin version.

@peppy Do you use windows as well? If not, could you please also provide logs?

**Reply by peppy** [reply]
The logs are in the google drive link. I am on macOS.

**Reply by AlexPl292** [reply]
@peppy Yes, got it, thanks. You have exactly the same problem, so the next EAP (today or tomorrow) should fix it all.

**Reply by AlexPl292** [reply]
The fix is available in 0.69.3 EAP. Thank you for your contribution!

**Reply by peppy** [reply]
Now I can't get out of insert mode. Or maybe the plugin just isn't working at all, not sure. Let me know if you need logs.

**Reply by AlexPl292** [reply]
Oh, troubles continue :(
Yes, the logs would be very appreciated.
Also, would it be possible to attach your .ideavimrc as a file? We've detected usages of `\r` symbol in your config (based on logs), but this sounds strange since you're using mac. Maybe there is something more unexpected.

**Reply by peppy** [reply]
I'll collect logs in a bit, but here's my configuration files in the mean time https://gist.github.com/peppy/87b30473d0b8c210d92fc92c4c15a22e.

I noticed line endings were set incorrectly on `.ideavimrc` (probably because I share this with a windows install) so I've corrected that. Have not tested again yet.

**Reply by AlexPl292** [reply]
New fixes based on your configs are released on 0.69.4 EAP.

**Reply by peppy** [reply]
Working well now, thanks

**Comment by momadacoding** [comment]
Hello, just found this, and it's exciting 👍 
I have the same issue with this link
https://youtrack.jetbrains.com/issue/VIM-788

The map command was found in vim wiki: https://vim.fandom.com/wiki/Searching#Case_sensitivity

Can you manage to make this work? please, I like this mapping

**Reply by lippfi** [reply]
Hi! We will definitely make it work, but in future. At the moment we are focused on some basic vimscript support like proper scopes behavior. This mapping requires the "expand" function and expression register to be implemented first

**Reply by SebastianDix** [reply]
Hello, will you ever implement "expand"? :-)

**Reply by lippfi** [reply]
@SebastianDix For multiple reasons, new functionality is unlikely to be added (at least by maintainers) in the immediate future:
- Our current focus is on various refactorings to support the new Fleet IDE. Although we are addressing bugs, adding new functionality isn't our priority at this time.
- The primary issue isn't the expand function, but the limitations in the search/command prompt. Features like the expression register (`<C-R>=`) and non-ASCII characters won't work within these limitations. Therefore, rewriting the prompts is a much higher priority and is more likely to be addressed in the near future.
- There are about 600 VimScript functions, and it's a very time-consuming task to implement them. So, while we have significant bugs to address, I don't want to take on this task as it's not the priority. Since IdeaVim is an open-source project, I'm hopeful that the community will contribute here. After all, implementing new functions is a relatively easy task.

**Comment by nov1n** [comment]
Hi! Congrats on this feature! The following line from my configuration no longer works.
```
inoremap jk <Esc>
```
Pressing jk simply types 'jk' on IdeaVim 1.8. I confirmed that downgrading to version 0.69 restores the functionality.

**Reply by lippfi** [reply]
Hi! Could you please provide your `.ideavimrc` and `.vimrc`? If you type `:inoremap jk <Esc>` in editor, then the mapping works fine, something must have went wrong during your config execution

**Reply by nov1n** [reply]
Hmm, you're right. If I type it in the editor it works for me too, but somehow it doesn't get picked up from the `.ideavimrc`. Here's the full config file. `.ideavimrc` is symlinked to `.vimrc`.

```
scriptencoding utf-8
set encoding=utf-8
set fileencoding=utf-8
set nocompatible                            " Only compile as vim
set t_Co=256                                " Support 256bit terminal colors
set hidden                                  " Allow buffer switch w/o saving
let $FZF_DEFAULT_COMMAND= 'ag --ignore vendor --ignore .git -g ""'

" Load vim-plug if not installed
if empty(glob("~/.vim/autoload/plug.vim"))
    execute '!curl -fLo ~/.vim/autoload/plug.vim https://raw.github.com/junegunn/vim-plug/master/plug.vim'
endif

call plug#begin('~/.vim/plugged')
Plug 'airblade/vim-gitgutter'             " Shows git edits in gutter
Plug 'bling/vim-airline'                  " Improved status bar
Plug 'ConradIrwin/vim-bracketed-paste'    " Handle paste mode before pasting with ctl-shift-v
Plug 'ervandew/supertab'                  " Autocomplete with tab
Plug 'fatih/vim-go'                       " Go support
Plug 'junegunn/fzf.vim'                   " Fuzzy file finding
Plug '/usr/local/opt/fzf'                 " Fzf executable
Plug 'junegunn/goyo.vim'                  " Distraction free writing
Plug 'klen/python-mode'                   " Do all python things
Plug 'maralla/completor.vim'              " Async autocompletion
Plug 'mileszs/ack.vim'                    " Searching
Plug 'moll/vim-bbye'                      " Sane buffer closing
" Plug 'morhetz/gruvbox'                    " Theme
Plug 'honza/vim-snippets'                 " Snippet library
Plug 'nelstrom/vim-visual-star-search'    " Use * to search visual selection
Plug 'nsf/gocode', { 'rtp': 'vim', 'do': '~/.vim/plugged/gocode/vim/symlink.sh' }
Plug 'raimondi/delimitmate'               " Auto close brackets etc
Plug 'scrooloose/nerdtree'                " File explorer
Plug 'sheerun/vim-polyglot'               " Language pack
Plug 'SirVer/ultisnips'                   " Snippet support
Plug 'tomtom/tcomment_vim'                " Add comment shortcuts
Plug 'tpope/vim-fugitive'                 " Git wrapper
Plug 'tpope/vim-rhubarb'                  " Github browsing with :Gbrowse
Plug 'tpope/vim-surround'                 " Surround with
Plug 'tpope/vim-unimpaired'               " Bracket mappings
Plug 'vim-airline/vim-airline-themes'     " Themes for airline
Plug 'connorholyday/vim-snazzy'           " Theme
Plug 'w0rp/ale'                           " Async linters
Plug 'iamcco/markdown-preview.nvim', { 'do': { -> mkdp#util#install() }}
Plug 'aserebryakov/vim-todo-lists'
call plug#end()

filetype plugin indent on                   " Expose filetype
syntax on                                   " Enable syntax highlighting
colorscheme snazzy

set number                                  " Show line numbers
set relativenumber                          " Show relative numbers
set incsearch                               " Search on keyup
set ignorecase                              " Case insensitive search
set smartcase                               " Be smart about case when searching
set hlsearch                                " Highlight all occurrences
set list                                    " Show tabs as |
" set listchars=tab:\ ,trail:·,eol:¬,nbsp:_
set listchars=tab:»\ ,extends:›,precedes:‹,nbsp:·,trail:·

set backspace=indent,eol,start              " Backspace behaves normally

nnoremap <C-J> <C-W><C-J>|                  " Easier navigation in splits
nnoremap <C-K> <C-W><C-K>
nnoremap <C-L> <C-W><C-L>
nnoremap <C-H> <C-W><C-H>

vnoremap <Tab> >|                           " Indent with tab in visual mode
vnoremap <S-Tab> <

noremap <leader>cc :TComment<CR>|           "Toggle comments
noremap <leader>cb :TCommentBlock<CR>|      "Toggle comments as a block

set splitbelow                              " Open new splits below and right
set splitright

cnoreabbrev Wq wq|                          " Correct common spelling mistakes
cnoreabbrev WQ wq
cnoreabbrev W w
cnoreabbrev Q q

nnoremap <silent><C-b> :<C-u>nohlsearch<CR><C-l>| " Remove highlighting
nnoremap <C-b> :nohl<CR>
nmap <leader>w :w!<cr>|                     " Use w! by default
set so=7                                    " Set 7 lines to the cursor

set ruler                                   " Show ruler

set showmatch                               " Show matching brackets
set mat=2                                   " Blink time

set noerrorbells                            " No annoying sound on errors
set novisualbell
set t_vb=
set tm=500

set background=dark                         " Use dark background

set ffs=unix,dos,mac                        " Use Unix as standard filetype

set noswapfile                              " Disable swap and backup measures set nowb
set nobackup

set expandtab                               " Use spaces instead of tabs
set smarttab                                " Be smart with tabs
set shiftwidth=4                            " 1 tab == 4 spaces
set tabstop=4
set autoindent                              " Set auto indent
set smartindent                             " Set smart indent
"set wrap                                    " Wrap lines

map j gj|                                   " Treat long lines as break lines
map k gk

set clipboard=unnamed                       " Alias the unnamed register to the + register (clipboard)

set laststatus=2                            " Always show status line

set guioptions-=m                           " Remove menu bar
set guioptions-=T                           " Remove toolbar
set guioptions-=r                           " Remove right-hand scroll bar
set guioptions-=L                           " Remove left-hand scroll bar

let delimitMate_expand_cr = 1 |             " Auto indent after bracket

inoremap jk <Esc>|                          " Experimental

                                            " Omnicomplete
inoremap <expr> <C-Space> pumvisible() \|\| &omnifunc == '' ?
            \ "\<lt>C-n>" :
            \ "\<lt>C-x>\<lt>C-o><c-r>=pumvisible() ?" .
            \ "\"\\<lt>c-n>\\<lt>c-p>\\<lt>c-n>\" :" .
            \ "\" \\<lt>bs>\\<lt>C-n>\"\<CR>"
imap <C-@> <C-Space>

let g:airline_powerline_fonts = 1
let g:airline_theme='wombat'
let g:airline#extensions#tabline#enabled = 1
let g:airline#extensions#ale#enabled = 1

" Automatically insert import paths
let g:go_fmt_command = "goimports"

" Exclude vendor directory from search
set wildignore+=*vendor/*

" Fix go save hangup, remove 'go' if slow
let g:syntastic_go_checkers = ['go', 'golint', 'govet', 'errcheck']
set statusline+=%#warningmsg#
set statusline+=%{SyntasticStatuslineFlag()}
set statusline+=%*

" Open nerdtree on open
"autocmd StdinReadPre * let s:std_in=1
"autocmd VimEnter * if argc() == 0 && !exists("s:std_in") | NERDTree | endif
map <C-n> :NERDTreeToggle<CR>

" Close nerdtree when it's the only window left
autocmd bufenter * if (winnr("$") == 1 && exists("b:NERDTree") && b:NERDTree.isTabTree()) | q | endif

" Folding
set foldmethod=syntax
set foldnestmax=1
set foldlevel=20

" Close buffers with \q
:nnoremap <Leader>q :Bdelete<CR>

" Make the tee hack easier
cmap w!! w !sudo tee > /dev/null %

" Go keybindings
"au FileType go nmap <leader>r <Plug>(go-run)
au FileType go nmap <leader>b <Plug>(go-build)
au FileType go nmap <leader>t <Plug>(go-test)
au FileType go nmap <leader>c <Plug>(go-coverage)

au FileType go nmap <leader>gk <Plug>(go-doc-browser)
au FileType go nmap <leader>gi <Plug>(go-info)
au FileType go nmap <leader>gr <Plug>(go-rename)
au FileType go nmap <leader>gm :GoImport

" Use gometalinter (combines different linters)
" let g:ale_linters = {'go': ['gometalinter']}

" Snippets
au FileType go imap <C-e> if err != nil {<CR>
let g:UltiSnipsJumpForwardTrigger="<c-f>"
let g:UltiSnipsJumpBackwardTrigger="<c-b>"

let g:SuperTabDefaultCompletionType = "context"

" Python PEP8
au BufNewFile,BufRead *.py
    \ set tabstop=4 |
    \ set softtabstop=4 |
    \ set shiftwidth=4 |
    \ set textwidth=79 |
    \ set expandtab |
    \ set autoindent |
    \ set fileformat=unix

" Plugins are sourced after this file which override mappings, the following
" commands are loaded after plugin rc files.
"
" These bindings are special and loaded after plugin to overwrite mappings
" defined by plugins
" https://vi.stackexchange.com/questions/756/how-can-i-redefine-plugin-key-mappings
au VimEnter * inoremap <S-Tab> <Esc>la
au VimEnter * nnoremap <buffer> <silent> <C-t> :Files<CR>
au FileType go nnoremap <buffer> <silent> <C-t> :Files<CR>

" Rerun last command in top right tmux pane
"                                  session:window.pane
nmap <leader>r :!tmux send-keys -t main:dev.1 C-p C-j<CR><CR>

" fzf keybindings
nmap , :Buffers<CR>
nmap <C-g> :Tags<CR>

" Prose mode
function! ProseMode()
  call goyo#execute(0, [])
  set spell noci nosi noai nolist noshowmode noshowcmd
  set complete+=s
  set bg=light
endfunction

command! ProseMode call ProseMode()
nmap \p :ProseMode<CR>

```

**Reply by lippfi** [reply]
Vimscript's grammar rule for comments was broken, sorry. The fix will be available in the next release. Thanks a lot for your report, it was very helpful

**Reply by nov1n** [reply]
Thanks for your quick turnaround. Happy to help and looking forward to the next release!

**Comment by momadacoding** [comment]
`vnoremap // y/\V<C-R>=escape(@",'/\')<CR><CR>`
How can I make this works ?  this map is copy from vimwiki https://vim.fandom.com/wiki/Search_for_visually_selected_text

**Reply by lippfi** [reply]
The escape function doesn't look difficult to implement (at least at the first sight), the main issue here is the expression evaluation in the search prompt `<C-R>=` which is not supported. Current implementation of the search prompt (and command prompt) is very limited at the moment

**Comment by SerkanSipahi** [comment]
@nov1n 

I would be very pleased if you could explain the reason for the pipe (|) here, thank you.

```vim
vnoremap <Tab> >|                           " Indent with tab in visual mode
...
...
noremap <leader>cc :TComment<CR>|           "Toggle comments
noremap <leader>cb :TCommentBlock<CR>|      "Toggle comments as a block
...
...
...
cnoreabbrev Wq wq|                          " Correct common spelling mistakes
...
```

**Reply by nov1n** [reply]
I honestly can't remember. I made this config a long time ago when I just started out with Vim, so it could very well be that the pipes do nothing.

I've since switched to Neovim, but I kept this config around for the Intellij vim emulation.

**Reply by lippfi** [reply]
See `:h comment`:
```It is not possible to add a comment to a shell command ":!cmd" or to the
":map" command and a few others (mainly commands that expect expressions)
that see the '"' as part of their argument:```

**Comment by vegerot** [comment]
Impressive!  Just wondering: have we ever considered running real Neovim, embedded in IntelliJ?  This is how the VSCode Neovim plugin works.  It would give us 100% vimscript and lua compatibility

**Reply by lippfi** [reply]
As a team of Vim lovers inside JetBrains who work on IdeaVim part-time, we really want to achieve the best Vim support possible, and embedding NeoVim seemed like a cool idea to consider at some point. I mean, it sounds easier than implementing things from scratch in Kotlin, so we indeed considered this option. I'm a NeoVim user myself and it's good; I don't have anything against it.
However, although embedding NeoVim can make some things better, like Vim Script support out of the box, it will introduce a bunch of other problems. 
More than that, we believe that in the long run, having a Kotlin Vim library that could be run in any IDE is the right direction. IdeaVim is now used in IJ Platform and Fleet, and it's much easier to integrate into various products because it's just a set of interfaces with bare Vim logic that could bring Vim to any editor. IdeaVim aims to be about tight IDE compatibility, and we are working to make it work everywhere, like settings, tool windows, and other non-editor components, unlike traditional Vim.
And yes, we may miss Lua and Vim Script, but I believe configurations in Kotlin could be a good alternative.
So, we have our vision and believe that having a Vim library that will be able to get the best of both worlds (IDE and Vim) is superior to bringing Vim into an IDE editor. But it will take some time to make this true.

**Comment by sharpchen** [comment]
Is `mode()` supported? Is there a full list of supported vim functions?
